### PR TITLE
[CALCITE-6674] Make RelDecorrelator rules configurable

### PR DESCRIPTION
**Pull request by rubenada**
See https://issues.apache.org/jira/browse/CALCITE-6674

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java [hunk 1]
```diff
@@ -354,26 +376,54 @@ protected RelNode decorrelate(RelNode root) {
...
+            context,
+            true,
+            createCopyHook(),
+            RelOptCostImpl.FACTORY);
+    planner.setDecorrelator(this);
+    return planner;
   }
 
+  /**
+   * Tries to remove the {@link org.apache.calcite.rel.core.Correlate} instances of a query
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:391]: I think "removes some instances of Correlate from a query plan" is a bit more precise.

**Review by mihaibudiu — APPROVED** [review]
I haven't checked that the decorrelator is always obtained from the planner (i.e., the replacement has been done everywhere in the codebase), but these changes look fine.

**Review by julianhyde — APPROVED** [review]
+1

The one change I would make is for `Planner.getDecorrelator()` to return not-null, and throw a message like "No decorrelator" if it has not been set.

**Review by caicancai — APPROVED** [review]

**Comment by rubenada** [comment]
Thanks for the feedback. I have addressed the suggestions.

**Comment by mihaibudiu** [comment]
You can probably squash the commits and merge this PR.

**Pull request merged**